### ISSUE TITLE
Credentials are logged when connection to PostGIS source fails using ogr2ogr 

### ISSUE THREAD

**Issue by leofuhrmann** [OSGeo/gdal#9318]
## Expected behavior and actual behavior.

When the connection to a Postgres/PostGIS database fails, I expect the logs to contain redacted credentials. This is the case when the database acts as ogr2ogr's destination (password is replaced by XXXX...). However, When the database is the source, the credentials are logged in plaintext.

## Steps to reproduce the problem.

Using ogr2ogr to create a GeoPackage (or any other file format) from a PostGIS database:

```sh
ogr2ogr -f GPKG output.gpkg PG:"dbname=example user=postgres password=shouldBeRedacted host=localhost port=5432"
```

When the connection fails (e. g., database does not exist or server not running), this results in

> ERROR 1: PQconnectdb failed.
> [...]
> FAILURE:
Unable to open datasource `PG:dbname=example user=postgres password=**shouldBeRedacted** host=localhost port=5432' with the following drivers.
> [...]

For comparison, if PostGIS is the destination, the password gets redacted as desired:

> ERROR 1: PostgreSQL driver failed to create PG:dbname=example user=postgres password=**XXXXXXXXXX** host=localhost port=5432

## Operating system

macOS Sonoma 14.3.1

## GDAL version and provenance

GDAL 3.8.3
Also tested using current Docker image ghcr.io/osgeo/gdal:alpine-small-3.8.4

**Comment by leofuhrmann** [comment]
This issue replaces #9317, which was opened under a wrong username

**Comment by leofuhrmann** [comment]
https://github.com/OSGeo/gdal/blob/master/port/cpl_error.cpp#L447 This seems to be the part where password redaction/obfuscation is happening. I'm inexperienced with c++, so I'm not able to work out a solution right now.

**Comment by leofuhrmann** [comment]
Thanks @rouault for quickly fixing the problem. I'm aware that 3.8.5 was just released, however the change is not included in this release. Can you tell, if the change will make it to the next release?

**Comment by rouault** [comment]
yes, this was a master-only change, so this will be included in GDAL 3.9.0 beginning of May
